internal/graph/check.go

```diff
@@ -614,6 +614,238 @@ func (c *LocalChecker) ResolveCheck(
+	if !ok {
+		return nil, fmt.Errorf("typesystem missing in context")
+	}
+
+	ds, ok := storage.RelationshipTupleReaderFromContext(ctx)
+	if !ok {
+		return nil, fmt.Errorf("relationship tuple reader datastore missing in context")
+	}
+
+	// usersetsMap is a map of all ObjectRelations and its IdsFor example,
+	// [group:1#member, group:2#member, group:1#owner, group:3#owner] will be stored as
+	// [group#member][1]
+	// [group#owner][1, 3]
+	usersetsMap := make(map[string]map[string]struct{})
+
+	for {
+		t, err := iter.Next(ctx)
+		if err != nil {
+			if errors.Is(err, storage.ErrIteratorDone) {
+				break
+			}
+
+			return nil, err
+		}
+
+		condEvalResult, err := eval.EvaluateTupleCondition(ctx, t, typesys, req.GetContext())
+		if err != nil {
+			errs = errors.Join(errs, err)
+			continue
+		}
+
+		if len(condEvalResult.MissingParameters) > 0 {
+			errs = errors.Join(errs, condition.NewEvaluationError(
+				t.GetCondition().GetName(),
+				fmt.Errorf("tuple '%s' is missing context parameters '%v'",
+					tuple.TupleKeyToString(t),
+					condEvalResult.MissingParameters),
+			))
+
+			continue
+		}
+
+		if !condEvalResult.ConditionMet {
+			continue
+		}
+
+		object, relation := tuple.SplitObjectRelation(t.GetUser())
+		objectType, objectID := tuple.SplitObject(object)
+		objectRel := tuple.ToObjectRelationString(objectType, relation)
+		if _, ok := usersetsMap[objectRel]; !ok {
+			usersetsMap[objectRel] = make(map[string]struct{})
+		}
+		if _, ok := usersetsMap[objectRel][objectID]; !ok {
+			usersetsMap[objectRel][objectID] = struct{}{}
+		}
+	}
+
+	if errs != nil {
+		telemetry.TraceError(span, errs)
+		return nil, errs
+	}
+
+	storeID := req.GetStoreID()
+	reqTupleKey := req.GetTupleKey()
+
+	// Next, for all the ObjectRelation, compare the associated objectIDs
+	// to the users associated objects
+	// all of this can likely bee its own function
+	handlers := make([]CheckHandlerFunc, 0, len(usersetsMap))
+	for objectRel, objectIDs := range usersetsMap {
+		handler := func(ctx context.Context) (*ResolveCheckResponse, error) {
+			response := &ResolveCheckResponse{
+				Allowed: false,
+				ResolutionMetadata: &ResolveCheckResponseMetadata{
+					DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount + 1,
+				},
+			}
+			objectType, relation := tuple.SplitObjectRelation(objectRel)
+			iter, err := ds.ReadStartingWithUser(ctx, storeID, storage.ReadStartingWithUserFilter{
+				ObjectType: objectType,
+				Relation:   relation,
+				UserFilter: []*openfgav1.ObjectRelation{{
+					Object: reqTupleKey.GetUser(),
+				}},
+			})
+			if err != nil {
+				return nil, err
+			}
+			defer iter.Stop()
+
+			for {
+				t, err := iter.Next(ctx)
+				if err != nil {
+					if errors.Is(err, storage.ErrIteratorDone) {
+						break
+					}
+
+					return nil, err
+				}
+
+				if actualObject := t.GetKey().GetObject(); actualObject != "" {
+					_, objectID := tuple.SplitObject(actualObject)
```

Comment:
FYI Updated godoc for this function: https://github.com/openfga/openfga/pull/1737